### ISSUE TITLE
Add group_by/3 and group_with/3

### ISSUE THREAD

**Issue by mndrix** [mndrix/list_util#7]
These are similar to Haskell's [groupBy](http://hackage.haskell.org/package/base-4.7.0.0/docs/Data-List.html#v:groupBy) and [groupWith](http://hackage.haskell.org/package/base-4.7.0.0/docs/GHC-Exts.html#v:groupWith).

These predicates have a natural semantic in reverse (`flatten/2`).  That predicate is deprecated so we should probably throw an exception rather than support bidirectional operation.  As long as we throw an exception, we can change our mind later about supporting other modes.

**Comment by mndrix** [comment]
group_with/3 was added in commit a7dd0a7260c65e505e68086fce0ac752bc8b3f91

**Comment by eazar001** [comment]
`group_by/3` added in 1533bd5.

**Comment by mndrix** [comment]
Reopening the issue while we discuss the semantics of commit 1533bd507c2f7cc671c47186fbfa6239c2087677.

**Comment by eazar001** [comment]
Hey @mndrix I read your arguments in favor of the Haskell semantics at 1533bd5, and they all seem very reasonable to me. So given your proposed adjustments it seems like we want something a little more like (with a little cleanup):

```
group_by(_, [], []) :- !.
group_by(_, [X], [[X]]) :- !.

group_by(Goal, [X,Y|Rest], [[X|Group]|Groups]) :-
    call(Goal, (=), X, Y), !,
    group_by(Goal, [Y|Rest], [Group|Groups]).

group_by(Goal, [X,Y|Rest], [[X]|Groups]) :-
    group_by(Goal, [Y|Rest], Groups).
```

I'll clean that up a little bit, push it in, and see if I can modify the tests a little later tonight.

**Comment by eazar001** [comment]
Actually, I just realized it's probably better to have something more like:
```
call(Goal, X, Y)
```
rather than:
```
call(Goal, Order, X, Y)
```
since `LT` and `GT` no longer make any sense with the new semantics.

**Comment by eazar001** [comment]
Actually, now I can't really decide here. Should we keep the comparison format of `call(Goal, Order, X, Y)` or use `call(Goal, X, Y)`. Personally the second case seems more direct to me, but I can't really see the benefit to using the first (aside from being able to utilize `compare/3`). 

I suppose I'll hold off on any changes until we can reach an agreement over this.

**Comment by mndrix** [comment]
> Should we keep the comparison format of call(Goal, Order, X, Y) or use call(Goal, X, Y)

I prefer `call(Goal,X,Y)` too.  It requires the caller to provide only the essential functionality that's needed for the predicate.  Users can still pass `==/2` as the `Goal` if they want grouping by the standard comparison order.

I suppose it wouldn't hurt to add `group/2` while you're at it:

```prolog
group(List,Groups) :-
    group_by(==,List,Groups).
```

**Comment by eazar001** [comment]
Alright, added in 5595f42b946d3e7141a1fd6db0804b4c80ccd669. Let me know if that works for you.

**Comment by mndrix** [comment]
That works great.  As I was reviewing the tests, I had an idea for a tail recursive version of the predicate so I implemented that in commit d57eec54680d10e5f40196e52e45be4e686217f6.

I'll include all these changes in the next release, probably tomorrow.  Thanks @eazar001 for your work on this.

**Comment by eazar001** [comment]
@mndrix , very nice catch on the backwards mode bug. A simple but nasty one ... invalid inputs would surely have passed otherwise.

**Comment by mndrix** [comment]
Released as part of v0.11.0
